backend/runner/runner.go

```diff
@@ -298,3 +312,53 @@ func (s *Service) registrationLoop(ctx context.Context, send func(request *ftlv1
 	}
 	return nil
 }
+
+func (s *Service) streamLogsLoop(ctx context.Context, send func(request *ftlv1.StreamDeploymentLogsRequest) error) error {
+	delay := time.Millisecond * 500
+
+	deployment, ok := s.deployment.Load().Get()
+	if !ok {
+		// No deployment, nothing to do.
+		select {
+		case <-ctx.Done():
+			err := context.Cause(ctx)
+			return errors.WithStack(err)
+		case <-time.After(delay):
+			return nil
+		}
+	}
+
+	select {
+	case entry := <-deployment.logQueue:
+		var errorString string
+		if entry.Error != nil {
+			errorString = entry.Error.Error()
+		}
+		err := send(&ftlv1.StreamDeploymentLogsRequest{
+			DeploymentKey: deployment.key.String(),
+			RunnerKey:     s.key.String(),
+			TimeStamp:     entry.Time.Unix(),
+			LogLevel:      int32(entry.Level.Severity()),
+			Attributes:    entry.Attributes,
+			Message:       entry.Message,
+			Error:         &errorString,
+		})
+		if err != nil {
+			return errors.WithStack(err)
+		}
+	case <-time.After(delay):
+	}
```

Comment:
only sending one per loop here I think. I could also do a `for` to stream them all but not 100% which approach is best.

Reply:
Sounds good. This select should be checking ctx.Done().